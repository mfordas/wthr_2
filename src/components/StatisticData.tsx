import React, {useMemo} from 'react';
import {StyleSheet, Text, View} from 'react-native';

import {constants, types} from 'app/constants';
import {
  prepareStatisticsData,
  countMaxValue,
  countMeanValue,
  countMinValue,
  countModeValue,
} from 'app/utils/statistics';

interface Props {
  fiveDaysForecast: types.apiOneDayForecastResponse[];
}

export const StatisticData: React.FC<Props> = ({fiveDaysForecast}) => {
  const dataForStatistics = useMemo(
    () => prepareStatisticsData(fiveDaysForecast),
    [fiveDaysForecast],
  );

  const maxValue: number = useMemo(
    () => countMaxValue(dataForStatistics.maxTemperatures),
    [dataForStatistics],
  );
  const minValue: number = useMemo(
    () => countMinValue(dataForStatistics.minTemperatures),
    [dataForStatistics],
  );
  const modeValue: number[] | false = useMemo(
    () =>
      countModeValue([
        ...dataForStatistics.minTemperatures,
        ...dataForStatistics.maxTemperatures,
      ]),
    [dataForStatistics],
  );

  const meanValue: number = useMemo(
    () =>
      countMeanValue([
        ...dataForStatistics.minTemperatures,
        ...dataForStatistics.maxTemperatures,
      ]),
    [dataForStatistics],
  );

  return (
    <View style={styles.container}>
      <Text style={styles.sectionText}>
        Max temperature {maxValue}
        {constants.CELSIUS_DEG}
      </Text>
      <Text style={styles.sectionText}>
        Min temperature {minValue}
        {constants.CELSIUS_DEG}
      </Text>
      {modeValue ? (
        <Text style={styles.sectionText}>
          Mode{modeValue.length > 1 ? 's' : ''} of temperatures{' '}
          {modeValue.map(mode => `${mode}${constants.CELSIUS_DEG} `)}
        </Text>
      ) : (
        <Text style={styles.sectionText}>There is not mode value</Text>
      )}
      <Text style={styles.sectionText}>
        Mean of temperatures {meanValue}
        {constants.CELSIUS_DEG}
      </Text>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: '90%',
    alignContent: 'center',
    alignItems: 'center',
  },
  sectionText: {
    fontSize: 20,
    paddingVertical: 12,
  },
});
