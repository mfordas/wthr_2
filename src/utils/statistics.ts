import {max, min, mean, mode, round} from 'mathjs';

import {types} from 'app/constants';

export const prepareStatisticsData = (
  apiData: types.apiOneDayForecastResponse[],
) => {
  const maxTemperatures = apiData.map(day => round(day.temp.max));
  const minTemperatures = apiData.map(day => round(day.temp.min));

  return {maxTemperatures, minTemperatures};
};

export const countMaxValue = (arrayOfValues: number[]) =>
  round(max(arrayOfValues));

export const countMinValue = (arrayOfValues: number[]) =>
  round(min(arrayOfValues));

export const countModeValue = (arrayOfValues: number[]) => {
  const modeOfNumbers: number[] = mode(arrayOfValues);
  return modeOfNumbers.length === arrayOfValues.length ? false : modeOfNumbers;
};

export const countMeanValue = (arrayOfValues: number[]) =>
  round(mean(arrayOfValues));
