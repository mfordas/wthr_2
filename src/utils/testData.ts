import {types} from 'app/constants';

export const mockedGeolocationApiCityReponse: types.apiGeocodeCityResponse = {
  name: 'London',
  lat: 12.123456789,
  lon: 12.123456789,
  local_names: [
    {
      af: 'Londen',
    },
  ],
  country: 'GB',
};

export const mockedOneDayForecastApiResponse: types.apiOneDayForecastResponse =
  {
    clouds: 52,
    dew_point: 12.27,
    dt: 1655373600,
    feels_like: {
      day: 24.53,
      eve: 18.1,
      morn: 11.91,
      night: 13.84,
    },
    humidity: 46,
    moon_phase: 0.58,
    moonrise: 1655415960,
    moonset: 1655352300,
    pop: 0.98,
    pressure: 1017,
    rain: 3.43,
    sunrise: 1655346977,
    sunset: 1655406530,
    temp: {
      day: 24.79,
      eve: 18,
      max: 26.05,
      min: 12.47,
      morn: 12.47,
      night: 14.13,
    },
    uvi: 5.09,
    weather: [
      {
        description: 'moderate rain',
        icon: '10d',
        id: 501,
        main: 'Rain',
      },
    ],
    wind_deg: 310,
    wind_speed: 6.65,
  };
