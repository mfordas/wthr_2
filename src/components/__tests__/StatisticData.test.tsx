import {render} from '@testing-library/react-native';
import React from 'react';

import {mockedOneDayForecastApiResponse} from 'app/utils/testData';

import {StatisticData, Props} from '../StatisticData';

const props: Props = {
  fiveDaysForecast: [mockedOneDayForecastApiResponse],
};
const wrapper = render(<StatisticData {...props} />);
test('StatisticData should match snapshot', () => {
  expect(wrapper).toMatchSnapshot();
});
