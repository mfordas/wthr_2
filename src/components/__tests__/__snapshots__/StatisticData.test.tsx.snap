// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticData should match snapshot 1`] = `
<View
  style={
    Object {
      "alignContent": "center",
      "alignItems": "center",
      "flex": 1,
      "width": "90%",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
        "paddingVertical": 12,
      }
    }
  >
    Max temperature 
    26
    °C
  </Text>
  <Text
    style={
      Object {
        "fontSize": 20,
        "paddingVertical": 12,
      }
    }
  >
    Min temperature 
    12
    °C
  </Text>
  <Text
    style={
      Object {
        "fontSize": 20,
        "paddingVertical": 12,
      }
    }
  >
    There is not mode value
  </Text>
  <Text
    style={
      Object {
        "fontSize": 20,
        "paddingVertical": 12,
      }
    }
  >
    Mean of temperatures 
    19
    °C
  </Text>
</View>
`;
